File: csharp/behavioral/classic_command.cs
namespace Gof.ModernCSharp.Behavioral;

// Command turns an action into an object that can be passed around and executed.
// Modern C# note: Action, Func<T>, and lambdas are often enough. Classes help
// when commands need undo, queues, metadata, serialization, or authorization.
public static class ClassicCommand
{
    sealed class Light { public string TurnOn() => "light on"; }
    interface ICommand { string Execute(); }
    sealed class LightOnCommand(Light light) : ICommand
    {
        public string Execute() => light.TurnOn();
    }
    sealed class Button(ICommand command)
    {
        public string Press() => command.Execute();
    }

    public static string Run() => new Button(new LightOnCommand(new Light())).Press();
}

File: py/behavioral/classic_command.py
# Command turns an action into an object that can be passed around and executed.
# Modern Python note: a function or closure is often enough. A class earns its keep
# when commands need metadata, undo, queuing, serialization, or authorization.
class Light:
    def turn_on(self):
        return "light on"

    def turn_off(self):
        return "light off"


class LightOnCommand:
    def __init__(self, light):
        self.light = light

    def execute(self):
        return self.light.turn_on()


class Button:
    def __init__(self, command):
        self.command = command

    def press(self):
        return self.command.execute()


if __name__ == "__main__":
    print(Button(LightOnCommand(Light())).press())
